File: api_tests/lib/asset.ts
import { Asset as AssetSdk } from "comit-sdk";
import { LedgerKind } from "./ledgers/ledger";

export interface Asset extends AssetSdk {
    name: AssetKind;
    ledger: LedgerKind;
    [k: string]: any;
}

export enum AssetKind {
    Bitcoin = "bitcoin",
    Ether = "ether",
    Erc20 = "erc20",
}

export function toKey(asset: Asset): string {
    return `${asset.name}-on-${asset.ledger}`;
}

export function toKind(key: string): { asset: AssetKind; ledger: LedgerKind } {
    switch (key) {
        case "bitcoin-on-bitcoin":
            return { asset: AssetKind.Bitcoin, ledger: LedgerKind.Bitcoin };
        case "bitcoin-on-lightning":
            return { asset: AssetKind.Bitcoin, ledger: LedgerKind.Lightning };
        case "ether-on-ethereum":
            return { asset: AssetKind.Ether, ledger: LedgerKind.Ethereum };
        case "erc20-on-ethereum":
            return { asset: AssetKind.Erc20, ledger: LedgerKind.Ethereum };
    }
}

File: cnd/src/asset.rs
mod bitcoin;
pub mod ethereum;
pub mod lightning;
pub use self::{
    bitcoin::Bitcoin,
    ethereum::{Erc20, Erc20Quantity, Ether},
    lightning::Lightning,
};
use crate::asset;
use derivative::Derivative;

#[derive(Clone, Derivative, PartialEq)]
#[derivative(Debug = "transparent")]
pub enum AssetKind {
    Bitcoin(Bitcoin),
    Ether(Ether),
    Erc20(Erc20),
}

impl From<Bitcoin> for AssetKind {
    fn from(amount: Bitcoin) -> Self {
        AssetKind::Bitcoin(amount)
    }
}

impl From<asset::Ether> for AssetKind {
    fn from(quantity: asset::Ether) -> Self {
        AssetKind::Ether(quantity)
    }
}

impl From<asset::Erc20> for AssetKind {
    fn from(quantity: asset::Erc20) -> Self {
        AssetKind::Erc20(quantity)
    }
}
